### PR TITLE
US027 validate collection instrument

### PR DESCRIPTION
**Pull request by andrewmil**
Added client side validation of the collection instrument file type.
Added error message boxes for invalid collection instrument types and for a failed upload of a collection instrument.

[Trello](https://trello.com/c/jyN3I43Y/58-us027-validate-collection-instrument-file-format-size-medium)
[Confluence](https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/230490136/US027+Validate+Collection+Instrument+s+Medium)

### CODE REVIEW

File: response_operations_ui/static/js/validate-ci.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		$("#ciFileErrorPanel").removeClass("panel--error");
+
+		$("#ciFileErrorPanelBody").removeClass("panel__body");
+		$("#ciFileErrorPanelBody p").addClass("hidden");
+
+		$("#btn-add-ci").removeClass("unready");
+
+	}else{
+
+		//alert("that's not an xlsx :(");
```
Comment (benjefferies) [inline response_operations_ui/static/js/validate-ci.js]: alert can be removed

File: response_operations_ui/static/js/validate-ci.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+function checkSelectedCI(files){
+
+	// Check for the various File API support.
+	if (window.FileReader) {
+		// FileReader are supported.
+		checkCI(files[0]);
+	} else {
+		alert('FileReader is not supported in this browser.');
```
Comment (benjefferies) [inline response_operations_ui/static/js/validate-ci.js]: Don't think we'd want to do this. We shouldn't do anything and let the server side do the rendering

File: response_operations_ui/templates/collection-exercise.html [hunk 3]
```diff
@@ -173,10 +175,17 @@ <h2 class="venus u-pt-m">Collection instruments</h2>
...
             <p>
-              {{ 'Add another collection instrument (CI)' if collection_instruments else 'Choose a collection instrument (CI) to load'}}
+              Add {{'another' if collection_instruments else 'a'}} collection instrument. Must be XLSX
             </p>
-            <label class="upload-file u-vh" for="ciFile">Choose CI file</label>
-            <input id="ciFile" type="file" class="" name="ciFile" accept=".xlsx">
+            <div id="ciFileErrorPanel" class="{{'panel panel--simple panel--error' if error.section == 'ciFile' else ''}}">
+              <div id="ciFileErrorPanelBody" class="{{'panel__body' if error.section == 'ciFile' else ''}}">
+                <p id="ciFileErrorText" class="hidden">
+                  Incorrect file type. Please choose a file type XLSX
```
Comment (jcox-dev) [inline response_operations_ui/templates/collection-exercise.html]: "Please choose a file *of* type XLSX"?

File: response_operations_ui/views/collection_exercise.py [hunk 4]
```diff
@@ -13,10 +13,11 @@
 
 
 @app.route('/surveys/<short_name>/<period>', methods=['GET'])
-def view_collection_exercise(short_name, period):
+def view_collection_exercise(short_name, period, error={}, ci_loaded=False, sample_loaded=False):
```
Comment (jcox-dev) [inline response_operations_ui/views/collection_exercise.py]: It's usually unsafe to have a mutable object (`error={}`) as a default parameter to a function as the same object can be used multiple times for multiple calls (created once at function definition time).

Maybe change to `error=None` and then in the return `error=error or {}` or `error={} if error is None else error` or similar.

**Comment by andrewmil** [comment]
[US027: Validate Collection Instrument (File Format) Size [Medium]?](https://trello.com/c/jyN3I43Y/58-us027-validate-collection-instrument-file-format-size-medium)

File: response_operations_ui/templates/collection-exercise.html (hunk 3 above)
Comment (andrewmil) [reply response_operations_ui/templates/collection-exercise.html]: The current text matches the designs, I'll pass it back by @nicprice first though

**Pull request merged**